### PR TITLE
Using bytebuffers instead of bytes

### PR DESCRIPTION
**Pull request by g4s8**
#52 - using `ByteBuffer` flow instead of `Byte` flow. Updated ASTO dependency, fixed all interfaces where needed, updated helper objects and adapters.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#53](https://github.com/artipie/docker-adapter/issues/53) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/2c1fd02f-21fa-48f9-b7f8-b8e9db57fa6e); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by fanifieiev** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/docker-adapter/issues/53#issuecomment-584362478))

@fanifieiev The user @fanifieiev[/z](https://www.0crat.com/u/fanifieiev) resigned from [#53](https://github.com/artipie/docker-adapter/issues/53), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/docker-adapter/issues/53#issuecomment-584362478))

@fanifieiev Job refused in 2 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
This pull request [#53](https://github.com/artipie/docker-adapter/issues/53) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3e90c385-73ec-4b9b-9012-5e03d6567caf); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by iakunin** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/docker-adapter/issues/53#issuecomment-584649262))

@iakunin Job refused in 16 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/docker-adapter/issues/53#issuecomment-584649262))

@iakunin The user @iakunin[/z](https://www.0crat.com/u/iakunin) resigned from [#53](https://github.com/artipie/docker-adapter/issues/53), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
This pull request [#53](https://github.com/artipie/docker-adapter/issues/53) is assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/695f6167-1118-44b4-8522-64299ceabf35); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/docker/misc/BytesFlowAs.java [hunk 1]
```diff
@@ -143,35 +140,50 @@ public void onError(final Throwable err) {
...
-            } catch (final Throwable err) {
+                this.future.complete(this.accum.value());
+            } catch (final IOException err) {
                 this.future.completeExceptionally(err);
             }
         }
     }
 
+    /**
+     * Accumulator for incomiing bytes chunks.
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/docker/misc/BytesFlowAs.java]: @g4s8 correct typo please in word "incoming"

File: src/main/java/com/artipie/docker/misc/BytesFlowAs.java [hunk 2]
```diff
@@ -195,8 +207,62 @@ public Text(final Flow.Publisher<Byte> flow) {
...
+         */
+        private final PipedInputStream inp;
+
+        /**
+         * Output pipe.
+         */
+        private final PipedOutputStream out;
+
+        /**
+         * Function to convert input stream tot target object.
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/docker/misc/BytesFlowAs.java]: @g4s8 correct typo here too please

File: src/main/java/com/artipie/docker/misc/BytesFlowAs.java [hunk 3]
```diff
@@ -195,8 +207,62 @@ public Text(final Flow.Publisher<Byte> flow) {
...
+        /**
+         * Ctor.
+         * @param func Accumulator function
+         */
+        @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
+        StreamAccum(final Function<InputStream, T> func) {
+            this.func = func;
+            this.out = new PipedOutputStream();
+            try {
+                this.inp = new PipedInputStream(this.out);
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/docker/misc/BytesFlowAs.java]: @g4s8 I'd suggest to extract this try-catch block into private static method and get rid of suppression

**Comment by g4s8** [comment]
@olenagerasimova thanks, please check new commit: https://github.com/artipie/docker-adapter/pull/53/commits/2173a057cb479b0f328af03e389755f6cccdd445

**Comment by g4s8** [comment]
@olenagerasimova I disabled a test for windows and added todo, can you please review it again?

**Comment by 0crat** [comment]
Job audit: performer @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)[/z](https://www.0crat.com/u/olenagerasimova) didn't make CR comments

**Comment by 0crat** [comment]
Job was finished in 37 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**